src/Transform/OptimizeMemoryPools.cpp

```diff
@@ -0,0 +1,905 @@
+
+using namespace mlir;
+namespace {
+
+// Handling of static memory pool on a block-basis in each function.
+typedef std::map<Block *, bool> BlockToCompactedFlag;
+std::map<FuncOp, std::unique_ptr<BlockToCompactedFlag>> staticPoolCompacted;
+
+/// Get the AllocOp of the current GetRef.
+AllocOp getAllocOfGetRef(KrnlGetRefOp *getRef) {
+  auto parentBlock = getRef->getOperation()->getBlock();
+
+  AllocOp alloc = nullptr;
+  parentBlock->walk([&alloc, getRef](AllocOp op) {
+    auto getRefAlloc = getRef->getOperands()[0];
+    if (op.getResult() == getRefAlloc)
+      alloc = op;
+  });
+
+  return alloc;
+}
+
+/// Get the number of GetRef ops associated with this AllocOp.
+int64_t getAllocGetRefNum(AllocOp *allocOp) {
+  auto parentBlock = allocOp->getOperation()->getBlock();
+
+  int64_t numGetRefs = 0;
+  parentBlock->walk([&numGetRefs, allocOp](KrnlGetRefOp op) {
+    auto result = allocOp->getResult();
+    if (op.getOperands()[0] == result)
+      numGetRefs++;
+  });
+
+  return numGetRefs;
+}
+
+/// Get the total size in bytes used by the getref operations associated
+/// with a given memory pool.
+int64_t getAllocGetRefTotalSize(AllocOp *allocOp) {
+  auto parentBlock = allocOp->getOperation()->getBlock();
+
+  int64_t totalSize = 0;
+  SmallVector<KrnlGetRefOp, 4> seenGetRefs;
+  parentBlock->walk([&totalSize, &seenGetRefs, allocOp](KrnlGetRefOp op) {
+    // If krnl.getref operation uses a different mempool then exit.
+    if (op.mempool() != allocOp->getResult())
+      return;
+
+    // Check that the krnl.getref operation has not already been counted.
+    // We must make sure we count the memory footprint of getref operations
+    // sharing a slot only once.
+    for (auto getRef : seenGetRefs)
+      if (op.offset() == getRef.offset())
+        return;
+
+    // Footprint has not been counted yet. Add it to totalSize.
+    // auto result = allocOp->getResult();
+    // if (op.getOperands()[0] == result)
+    totalSize += getMemRefSizeInBytes(op.getResult());
+
+    // Act krnl.getref operation as seen.
+    seenGetRefs.emplace_back(op);
+  });
+
+  return totalSize;
+}
+
+/// Returns a list of operations in the current block that use the getref.
+std::vector<Operation *> getGetRefStores(KrnlGetRefOp *getRef) {
+  auto parentBlock = getRef->getOperation()->getBlock();
+  std::vector<Operation *> stores;
+
+  parentBlock->walk([&stores, getRef](StoreOp op) {
+    for (const auto &operand : op.getOperands())
+      if (operand == getRef->getResult())
+        stores.emplace_back(op);
+  });
+
+  parentBlock->walk([&stores, getRef](AffineStoreOp op) {
+    for (const auto &operand : op.getOperands())
+      if (operand == getRef->getResult())
+        stores.emplace_back(op);
+  });
+
+  // The list contains at least one use.
+  return stores;
+}
+
+/// Returns a list of distinct krnl.getref operations in the current
+/// block that use the memory pool.
+SmallVector<KrnlGetRefOp, 4> getAllDistinctGetRefsForAlloc(AllocOp *allocOp) {
```

Comment:
The preferred class is SmallVectorImpl, see https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

It say that the Impl version is not really an implementation in the traditional sense and is compatible with the SmallVect... 

Though in this case, you return it... so it might be fine. Still, look at the page and make an opinion for yourself... It applies equally to the other usage of this same pattern below.